### PR TITLE
chore(docs): Use parser='mdx' for formatting docs using prettier

### LINKED ISSUE #21630: chore(docs): Use parser="mdx" for formatting docs
## Description

All our docs are read as MDX in our site, but because they all use the extension `.md`, they are treated as Markdown when formatted with prettier. This can lead to issues when prettier interprets code as text: #21609.

The fix is to (1) convert all our docs to use the `.mdx` file extension (bad idea) or (2), add a config override in our `.prettierrc.js`:

```
{
  "overrides": [
    {
      "files": "/docs/*.md",
      "options": {
        "parser": "mdx"
      }
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by akshitkrnagpal**
## Description

All Gatsby docs are read as MDX but the file extension is `.md`. Therefore prettier formats them as markdown and not mdx. This overrides the default behaviour for `.md` files and tells prettier to format them as mdx.

## Related Issues

Fixes #21630

### CODE REVIEW

File: docs/blog/2019-06-25-profitable-agency-wordpress/index.md [hunk 1]
```diff
@@ -9,7 +9,7 @@ tags:
   - aws
 ---
 
-import { Link } from "gatsby"
+import { Link as GatsbyLink } from "gatsby"
```
Comment (akshitkrnagpal) [inline docs/blog/2019-06-25-profitable-agency-wordpress/index.md]: This change is made to get around for this issue https://github.com/prettier/prettier/issues/7548.

File: docs/blog/2018-08-09-swag-store/index.md [hunk 2]
```diff
@@ -18,15 +18,14 @@ We’ve spent the last couple weeks working the kinks out, and we couldn’t be
     src="./images/gatsby-swag-twitter.jpg"
   />
   <figcaption>
-    Tweets from{" "}
-    <a href="https://twitter.com/DeaNHtiD99/status/1023204484183416832">Dean</a>,{" "}
-    <a href="https://twitter.com/ARebelBelle/status/1020044426712735744">
+    {/* prettier-ignore */}
```
Comment (tesseralis) [inline docs/blog/2018-08-09-swag-store/index.md]: hmm, what happens here if we don't put the prettier-ignore?

File: docs/blog/2019-06-25-profitable-agency-wordpress/index.md (hunk 1 above)
Comment (tesseralis) [inline docs/blog/2019-06-25-profitable-agency-wordpress/index.md]: why was this changed to GatsbyLink?

File: docs/docs/caching.md [hunk 3]
```diff
@@ -51,7 +51,12 @@ When deploying with Now, follow the instructions in the [Now documentation](http
 
 ---
 
-<sup>1</sup> You can use 'no-cache' instead of 'max-age=0, must-revalidate'. Despite what the name might imply, 'no-cache' permits a cache to serve cached content as long as it validates the cache freshness first. <sup>[2][3] </sup> In either case, clients have to make a round trip to the origin server on each request. However, if you are correctly utilizing ETags or Last-Modified validation you will avoid downloading assets when the cached copy is still valid (e.g. the file hasn't changed on the origin server since it was cached).
+<sup>
```
Comment (tesseralis) [inline docs/docs/caching.md]: This one's kind of obnoxious... could we ignore prettier for this paragraph?

File: docs/blog/2018-08-09-swag-store/index.md (hunk 2 above)
Comment (akshitkrnagpal) [reply docs/blog/2018-08-09-swag-store/index.md]: Prettier is unable to fix code style issues with this block.
`npx prettier --write docs/blog/2018-08-09-swag-store/index.md` always switches this block in these two versions but `--check` always fails.
<details>
<summary>Versions</summary>

Version 1
``` 
    Tweets from{" "}
    <a href="https://twitter.com/DeaNHtiD99/status/1023204484183416832">Dean</a>
    ,{" "}
    <a href="https://twitter.com/ARebelBelle/status/1020044426712735744">
      Elle
    </a>
    ,{" "}
    <a href="https://twitter.com/chaseadamsio/status/1021896138503245824">
      Chase
    </a>
    , and <a href="https://twitter.com/kato_katherine/status/1021967765400211456">
      Katherine
    </a> showing off their new Gatsby swag.
```

Version 2
```
    Tweets from{" "}
    <a href="https://twitter.com/DeaNHtiD99/status/1023204484183416832">Dean</a>,{" "}
    <a href="https://twitter.com/ARebelBelle/status/1020044426712735744">
      Elle
    </a>,{" "}
    <a href="https://twitter.com/chaseadamsio/status/1021896138503245824">
      Chase
    </a>
    , and <a href="https://twitter.com/kato_katherine/status/1021967765400211456">
      Katherine
    </a> showing off their new Gatsby swag.
```
</details>

File: docs/blog/2019-06-25-profitable-agency-wordpress/index.md (hunk 1 above)
Comment (akshitkrnagpal) [reply docs/blog/2019-06-25-profitable-agency-wordpress/index.md]: I changed this to get around the issue https://github.com/prettier/prettier/issues/7548

<Link> </Link> was being treated as <link /> (head tag) which cannot have children.

File: docs/docs/caching.md (hunk 3 above)
Comment (akshitkrnagpal) [reply docs/docs/caching.md]: Yeah. Sure. I'll update this.

Comment (tesseralis) [reply docs/docs/caching.md]: Hmm, I don't think I care that much. This whole page is weird: https://www.gatsbyjs.org/docs/caching/

1. There are no [2] or [3] references in the text
2. The footnotes don't have any links at all, making them annoying to reach.

File: docs/blog/2018-08-09-swag-store/index.md (hunk 2 above)
Comment (tesseralis) [reply docs/blog/2018-08-09-swag-store/index.md]: Hmmm that's annoying. it sucks that MDX can't treat inner content as markdown. I'll talk to the team about it and see if we can come up with a solution but otherwise I guess we'll have to leave it at this.

Comment (tesseralis) [reply docs/blog/2018-08-09-swag-store/index.md]: Okay apparently there _is_ a way to do it, by leaving a blank line in between the tags:

```
<Button>

Here is a [button](https://github.com)

</Button>
```

Unfortunately this seems to break Prettier's parsing:

https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeAZgSwOYFcBOcAfADpRkASchABJgM40CGNA2gDaZQDWAugBQB6AEZNRogJRkyqQVjyEiIADQgIABxiZo9ZKCb58EAO4AFAwl0omANwiYAJipDD8TMNzgwAyuvddsZBh8XDhVAAsYAFt2AHVwzHh6PzA4b0tEzBtEgE9kcHpdVS56ahhTN2wopmR0JnZS1QAregAPACE3Dy9vJii4ABkuOFr6xpAW1u8A9jgARVwIeFGGsJA-fFL8fKiHVud1fC4YWMcYcOQADgAGVUOIUti3dXyoOBtqZwBHRfgKjSsICY9AAtG84A4Ic5CD9MIQKkwqjUkHVVqpSlFMEEQmt6DN5r8RiixmsYGJTg5zsgAEyqYJMTCcKDYADCECi1Ve0BGqlwpQAKmIrKjxjZQgBJKCQ2DeMBHTQAQSl3hgOVmK1KAF9NUA

**Comment by akshitkrnagpal** [comment]
@tesseralis 
Can you check the failed test? Is it a flaky test?

**Comment by tesseralis** [comment]
Yeah, it’s due to this: https://github.com/gatsbyjs/gatsby/pull/21709

**Comment by akshitkrnagpal** [comment]
Hi @tesseralis 
Re-requesting a review. Let me know if something needs to change.

File: docs/docs/caching.md [hunk 4]
```diff
@@ -51,7 +51,9 @@ When deploying with Now, follow the instructions in the [Now documentation](http
 
 ---
 
+ <!-- prettier-ignore-start -->
```
Comment (wardpeet) [inline docs/docs/caching.md]: just checking why was this needed?

File: docs/blog/2019-06-25-profitable-agency-wordpress/index.md (hunk 1 above)
Comment (wardpeet) [reply docs/blog/2019-06-25-profitable-agency-wordpress/index.md]: Can we add prettier ignore instead? I feel we'll lose context on why this is here perhaps add a comment to the prettier issue.

File: docs/docs/caching.md (hunk 4 above)
Comment (akshitkrnagpal) [reply docs/docs/caching.md]: The changes in this block were not great and @tesserails suggested this change.
https://github.com/gatsbyjs/gatsby/pull/21671#discussion_r383638736

File: docs/blog/2019-06-25-profitable-agency-wordpress/index.md (hunk 1 above)
Comment (akshitkrnagpal) [reply docs/blog/2019-06-25-profitable-agency-wordpress/index.md]: I can add a comment above this. But do we really need that context?
Using `Link` from gatsby like this is still correct and we wouldn't really need to do anything after we bump prettier with a version with a fix for this. Adding a comment or `prettier-ignore` would require an additional step after the `prettier` bump.

Just a thought. Would still happily update the PR. What is your opinion?

Comment (wardpeet) [reply docs/blog/2019-06-25-profitable-agency-wordpress/index.md]: > I can add a comment above this. But do we really need that context?

I think we should, workarounds should always tell us why this is needed (a comment is fine). It seems a bit overkill for this one because it's already fixed but it's all about consistency. I can imagine someone stumble on this line of code and having to think about why we did `import { Link as GatsbyLink } from "gatsby"` and not `import { Link } from "gatsby"`

Comment (akshitkrnagpal) [reply docs/blog/2019-06-25-profitable-agency-wordpress/index.md]: Okay. Updated.

**Comment by akshitkrnagpal** [comment]
@wardpeet 
Can you check the failed test?

**Comment by tesseralis** [comment]
@akshitkrnagpal Nah, you're good. It's a flaky test that's failing due to npm audits. @wardpeet this should be good to go.

**Comment by wardpeet** [comment]
Thank you @akshitkrnagpal

**Comment by muescha** [comment]
@tesseralis i see some drawbacks with this PR

i remember there was a rule (and some PR) to remove line breaks to make translations more easy?


but in this PRR some childs are getting new line breaks:
- `Pullquote`
- `VisuallyHidden`

maybe there is a rule for prettier to leave (or change) text child nodes as one lines?

**Comment by muescha** [comment]
_PS: i can not find this one-line-rule in https://www.gatsbyjs.org/contributing/gatsby-style-guide/_
:(

**Comment by muescha** [comment]
i see you mention this with `hard wraps prose` in:
- #21743 `chore(www) Update nested markdown when prettier is upgraded to 1.19.2/1.20.0`
